File: paddle/tcmpt/core/dtype.cc
/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/tcmpt/core/dtype.h"

namespace pt {

std::ostream& operator<<(std::ostream& os, DataType dtype) {
  switch (dtype) {
    case DataType::kUndef:
      os << "Undefined";
      break;
    case DataType::kBOOL:
      os << "bool";
      break;
    case DataType::kINT8:
      os << "int8";
      break;
    case DataType::kUINT8:
      os << "uint8";
      break;
    case DataType::kINT16:
      os << "int16";
      break;
    case DataType::kINT32:
      os << "int32";
      break;
    case DataType::kINT64:
      os << "int64";
      break;
    case DataType::kFLOAT16:
      os << "float16";
      break;
    case DataType::kFLOAT32:
      os << "float32";
      break;
    case DataType::kFLOAT64:
      os << "float64";
      break;
    case DataType::kCOMPLEX64:
      os << "complex64";
      break;
    case DataType::kCOMPLEX128:
      os << "complex128";
      break;
    default:
      // TODO(chenweihang): change to enforce later
      throw std::runtime_error("Invalid DataType type.");
  }
  return os;
}

}  // namespace pt

File: python/paddle/_typing/dtype.py
from typing import Any, Literal, Type, TypeAlias, Union

import numpy as np


class dtype:
    def __init__(self, arg0: int) -> None: ...


uint8: dtype
int8: dtype
int16: dtype
int32: dtype
int64: dtype
float32: dtype
float64: dtype
float16: dtype
bfloat16: dtype
complex64: dtype
complex128: dtype
bool: dtype

_DTypeLiteral: TypeAlias = Literal[
    "uint8",
    "int8",
    "int16",
    "int32",
    "int64",
    "float32",
    "float64",
    "float16",
    "bfloat16",
    "complex64",
    "complex128",
    "bool",
]

_DTypeNumpy: TypeAlias = Union[
    Type[
        Union[
            np.uint8,
            np.int8,
            np.int16,
            np.int32,
            np.int64,
            np.float16,
            np.float32,
            np.float64,
            np.complex64,
            np.complex128,
            np.bool_,
        ]
    ],
    np.dtype[Any],
]

DTypeLike: TypeAlias = Union[dtype, _DTypeNumpy, _DTypeLiteral]
